src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.cs

```diff
@@ -0,0 +1,123 @@
+            Func<Stream, TJsonOptions, CancellationToken, ValueTask<TValue?>> deserializeMethod,
+            TJsonOptions jsonOptions,
+            CancellationToken cancellationToken)
+        {
+            if (client is null)
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
+
+            TimeSpan timeout = client.Timeout;
+
+            // Create the CTS before the initial SendAsync so that the SendAsync counts against the timeout.
+            CancellationTokenSource? linkedCTS = null;
+            if (timeout != Timeout.InfiniteTimeSpan)
+            {
+                linkedCTS = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                linkedCTS.CancelAfter(timeout);
+            }
+
+            // We call SendAsync outside of the async Core method to propagate exception even without awaiting the returned task.
+            Task<HttpResponseMessage> responseTask;
+            try
+            {
+                // Intentionally using cancellationToken instead of the linked one here as HttpClient will enforce the Timeout on its own for this part
+                responseTask = getMethod(client, requestUri, cancellationToken);
+            }
+            catch
+            {
+                linkedCTS?.Dispose();
+                throw;
+            }
+
+            return Core(client, responseTask, linkedCTS, deserializeMethod, jsonOptions, cancellationToken);
+
+            static async Task<TValue?> Core(
+                HttpClient client,
+                Task<HttpResponseMessage> responseTask,
+                CancellationTokenSource? linkedCTS,
+                Func<Stream, TJsonOptions, CancellationToken, ValueTask<TValue?>> deserializeMethod,
+                TJsonOptions jsonOptions,
+                CancellationToken cancellationToken)
+            {
+                try
+                {
+                    using HttpResponseMessage response = await responseTask.ConfigureAwait(false);
+                    response.EnsureSuccessStatusCode();
+
+                    Debug.Assert(client.MaxResponseContentBufferSize is > 0 and <= int.MaxValue);
+                    int contentLengthLimit = (int)client.MaxResponseContentBufferSize;
+
+                    HttpContent content = response.Content ?? throw new ArgumentNullException(nameof(content));
+
+                    if (content.Headers.ContentLength is long contentLength && contentLength > contentLengthLimit)
+                    {
+                        LengthLimitReadStream.ThrowExceededBufferLimit(contentLengthLimit);
+                    }
+
+                    try
+                    {
+                        using Stream contentStream = await HttpContentJsonExtensions.GetContentStreamAsync(content, linkedCTS?.Token ?? cancellationToken).ConfigureAwait(false);
+
+                        var lengthLimitStream = new LengthLimitReadStream(contentStream, (int)client.MaxResponseContentBufferSize);
```

Comment:
Can we avoid allocating this wrapper stream if MaxResponseContentBufferSize is the default int.MaxValue?

Reply:
Sure.
Looks like the `DeleteAsync` overload doesn't use `ResponseHeadersRead`, so we can avoid all of this overhead there as well.

Reply:
Looking at this again, I don't think we should avoid this wrapper - we'd still want to prevent the deserialize from reading more than 2 GB in that case.

Reply:
> The per-call allocation overhead disappears with zero-byte reads.

How so?

Reply:
The read in the wrapper stream is implemented like so:
```c#
ValueTask<int> readTask = _innerStream.ReadAsync(buffer, cancellationToken);

if (buffer.IsEmpty)
{
    return readTask;
}

if (readTask.IsCompletedSuccessfully)
{
    int read = readTask.Result;
    CheckLengthLimit(read);
    return new ValueTask<int>(read);
}

return Core(readTask); // Allocate
```

Assuming the caller is using zero-byte reads, you will alternate between
- The `buffer.IsEmpty` branch for the zero-byte read where we'll just pass through the task as we're not interested in the result (it won't affect the read limit)
- The `readTask.IsCompletedSuccessfully` branch after a zero-byte read as data is available immediately

You'll only hit the allocating path if the underlying stream doesn't support zero-byte reads, or if you hit an exception.

The [WebSocket telemetry stream](https://github.com/microsoft/reverse-proxy/blob/e46afb9a7db7e95977ae1c20b2e4d025173aa93c/src/ReverseProxy/WebSocketsTelemetry/WebSocketsTelemetryStream.cs#L74-L88) I mentioned is using exactly the same pattern. The result of the zero-byte read doesn't affect us as there isn't any data to parse.